src/cassandra_server.py

```diff
@@ -0,0 +1,202 @@
+
+"""Cassandra Server Object."""
+
+import contextlib
+import logging
+import secrets
+import string
+from typing import List, Optional
+
+from cassandra import ConsistencyLevel, InvalidRequest
+from cassandra.auth import PlainTextAuthProvider
+from cassandra.cluster import (
+    EXEC_PROFILE_DEFAULT,
+    Cluster,
+    ExecutionProfile,
+    NoHostAvailable,
+    Session,
+)
+from cassandra.policies import RoundRobinPolicy
+from cassandra.query import SimpleStatement
+from ops.charm import CharmBase
+from ops.framework import EventBase
+from ops.model import MaintenanceStatus
+
+CQL_PORT = 9042
+CQL_PROTOCOL_VERSION = 4
+ROOT_USER = "charm_root"
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterNotReadyError(Exception):
+    """Exception to raise when the cluster is not ready."""
+
+    pass
+
+
+class Cassandra:
+    """Class to manage connections to the Cassandra cluster."""
+
+    def __init__(self, charm: CharmBase):
+        """Cassandra class constructor.
+
+        Args:
+            charm: The charm that is using this class
+        """
+        self.charm = charm
+
+    @contextlib.contextmanager
+    def connect(
+        self, event: EventBase, username: str = ROOT_USER, password: Optional[str] = None
+    ) -> Session:
+        """Context manager to connect to the Cassandra cluster and return an active session.
+
+        Args:
+            event: The current event object
+            username: username to connect with
+            password: password to connect with
+
+        Returns:
+            A cassandra session
+        """
+        if password is None:
+            password = self.root_password(event)
+        auth_provider = PlainTextAuthProvider(username=username, password=password)
+        profile = ExecutionProfile(load_balancing_policy=RoundRobinPolicy())
+        cluster = Cluster(
+            [self.charm._hostname()],
+            port=CQL_PORT,
+            auth_provider=auth_provider,
+            execution_profiles={EXEC_PROFILE_DEFAULT: profile},
+            protocol_version=CQL_PROTOCOL_VERSION,
+        )
+        try:
+            session = cluster.connect()
+            yield session
+        finally:
+            cluster.shutdown
+
+    def create_user(self, event: EventBase, username: str) -> List[str]:
+        """Create a new Cassandra user.
+
+        Args:
+            event: The current event object
+            username: The username of the new user
+
+        Returns:
+            A list of the form [username, password]
+
+        """
+        password = self._generate_password()
+        try:
+            with self.connect(event) as conn:
+                conn.execute(
+                    f"CREATE ROLE IF NOT EXISTS '{username}' WITH PASSWORD = '{password}' AND LOGIN = true"
+                )
+        except NoHostAvailable as e:
+            logger.info("Caught exception %s:%s: deferring", type(e), e)
+            self.charm.unit.status = MaintenanceStatus("Waiting for Database")
```

Comment:
```suggestion
            self.charm.unit.status = MaintenanceStatus("Waiting for database")
```

Applies elsewhere too